student_and_courses/src/main/java/org/example/CourseDAO.java

```diff
@@ -0,0 +1,41 @@
+package org.example;
+
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.Persistence;
+
+public class CourseDAO {
+    private final EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory("jpa_practice-pu");
+
+    void addCourse(Course course) {
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        entityManager.getTransaction().begin();
+        entityManager.persist(course);
+        entityManager.getTransaction().commit();
+        entityManager.close();
+    }
```

Comment:
I see what you're doing and it's good you're using transactions - but your solution lacks robustness. Generally, if one part of the transaction fails, you want to do rollbacks. the best way to do it is to wrap it in a try-catch-finally statement, sth like this:
`    try {
        tx.begin();
        entityManager.persist(student);
        tx.commit();
    } catch (RuntimeException e) {
        if (tx.isActive()) tx.rollback();
        throw e; 
    } finally {
        entityManager.close();
    }`